### PR TITLE
Refactor core to use `e2e_invoke` host module.

### PR DESCRIPTION
**Pull request by dmkozh**
# Description

Refactor core to use `e2e_invoke` host module.

This simplifies the 'glue' layer between Core and host a bit and (more importantly) uses shared code that can also be re-used by the preflight.

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/InvokeHostFunctionOpFrame.cpp [hunk 1]
```diff
@@ -700,20 +593,24 @@ InvokeHostFunctionOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx,
...
-    int64_t rentFee = rust_bridge::compute_rent_fee(
-        cfg.CURRENT_LEDGER_PROTOCOL_VERSION,
-        ltx.loadHeader().current().ledgerVersion, rustEntryRentChanges,
-        sorobanConfig.rustBridgeRentFeeConfiguration(), ledgerSeq);
-    mParentTx.consumeRefundableSorobanResources(metrics.mEmitEventByte,
-                                                rentFee);
+
+    if (!mParentTx.consumeRefundableSorobanResources(
+            metrics.mEmitEventByte, out.rent_fee,
+            cfg.CURRENT_LEDGER_PROTOCOL_VERSION, sorobanConfig, cfg))
```
Comment (sisuresh) [inline src/transactions/InvokeHostFunctionOpFrame.cpp:599]: This should be passing the current protocol version (`ltx.loadHeader().current().ledgerVersion`), not `cfg.CURRENT_LEDGER_PROTOCOL_VERSION` from what I can see. `TransactionFrame::computeSorobanResourceFee` calls `rust_bridge::compute_transaction_resource_fee(
        cfg.CURRENT_LEDGER_PROTOCOL_VERSION, protocolVersion, ...);`, so you'd be passing the max in place of the current.

This comment applies to the calls in the other ops as well.

Comment (dmkozh) [reply src/transactions/InvokeHostFunctionOpFrame.cpp]: Nice catch, fixed.

**Comment by sisuresh** [comment]
r+ 91d50744a124b85083bc0293a2e7a76b1d2eb9f4

File: src/transactions/RestoreFootprintOpFrame.cpp [hunk 2]
```diff
@@ -123,15 +123,24 @@ RestoreFootprintOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx,
...
             .rustBridgeRentFeeConfiguration(),
         ledgerSeq);
-    mParentTx.consumeRefundableSorobanResources(metrics.mLedgerReadByte * 2,
-                                                rentFee);
+    if (!mParentTx.consumeRefundableSorobanResources(
+            0, rentFee, ltx.loadHeader().current().ledgerVersion,
+            app.getLedgerManager().getSorobanNetworkConfig(ltx),
+            app.getConfig()))
+    {
+        // TODO: This probably should have a more precise error code as here
```
Comment (sisuresh) [inline src/transactions/RestoreFootprintOpFrame.cpp:139]: Is there an issue for this?

Comment (dmkozh) [reply src/transactions/RestoreFootprintOpFrame.cpp:139]: No, I don't think so...

Comment (dmkozh) [reply src/transactions/RestoreFootprintOpFrame.cpp:139]: Opened https://github.com/stellar/stellar-core/issues/3868

**Pull request merged**